### PR TITLE
#11 - strict public method parameter checking

### PR DESCRIPTION
**Pull request by Ocramius**
Apologies for the branch name: I'm sleepy.

This PR introduces following behavior (providing implementation for #11)

```php
class Bar {}
class Foo {
    /** @param $foo \Bar[] */
    public function doFoo(array $foo) {}
}
```

This will work:

```php
$foo = new Foo();
$foo->doFoo([]);
$foo->doFoo([new Bar, new Bar]);
```

This will fail:

```php
$foo = new Foo();
$foo->doFoo([123, 456]); // crash
```

Tests and SoC are still to be addressed, this is just a PoC.

@malukenho please review :D

## TODOs:

 - [x] documentation
 - [x] unit tests
 - [x] remove code duplication
 - [x] proper dependency injection

### CODE REVIEW

File: tests/functional/call-to-method-with-typed-array-violation.phpt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    ->method([new \StrictPhpTestAsset\ClassWithTypedArrayMethodParameterAnnotation()]);
+
+echo "Never reached\n";
+
+?>
+--EXPECTF--
+OK
+OK
+OK
+%Aatal error: Argument 1 passed to %s must be an instance of stdClass, instance of %a given%a
```
Comment (malukenho) [inline tests/functional/call-to-method-with-typed-array-violation.phpt]: Missing `F` char?

Comment (Ocramius) [inline tests/functional/call-to-method-with-typed-array-violation.phpt]: Intentional. Depending on the PHP version, the message starts with "Fatal error" or "PHP fatal error"

Comment (malukenho) [inline tests/functional/call-to-method-with-typed-array-violation.phpt]: @Ocramius Oh! great!

File: src/StrictPhp/AccessChecker/ParameterTypeChecker.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            );
+        }
+    }
+
+    /**
+     * @param string                  $contextClass
+     * @param \ReflectionParameter[]  $reflectionParameters
+     * @param int                     $index
+     *
+     * @return Type[]
```
Comment (malukenho) [inline src/StrictPhp/AccessChecker/ParameterTypeChecker.php]: `@return Type[]|array`

Comment (Ocramius) [inline src/StrictPhp/AccessChecker/ParameterTypeChecker.php]: @malukenho I don't want to type-hint `array` here, as it's too loose

File: src/StrictPhp/TypeFinder/ParameterTypeFinder.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+use phpDocumentor\Reflection\Types\Static_;
+use ReflectionParameter;
+
+final class ParameterTypeFinder
+{
+    /**
+     * @param ReflectionParameter $reflectionParameter
+     * @param string              $contextClass
+     *
+     * @return \phpDocumentor\Reflection\Type[]
```
Comment (malukenho) [inline src/StrictPhp/TypeFinder/ParameterTypeFinder.php]: just `Type[]`

**Comment by malukenho** [comment]
:beers:

**Pull request merged**